datalad_next/runners/run.py

```diff
@@ -0,0 +1,258 @@
+    `arm`, and logic inside the timeout handler that can trigger a termination
+    or kill signal to the subprocess, if the termination time or kill time has
+    come.
+
+    Parameters
+    ----------
+    cls : type[Protocol]
+        A protocol class that should be extended by the kill-interface
+
+    Returns
+    -------
+    Any
+        If the `cls` is not a subclass of `GeneratorMixIn`, the result of
+        `cls._prepare_result` will be returned.
+
+    Generator
+        If the protocol is a subclass of `GeneratorMixIn`, the `Generator` that
+        is a result of `ThreadedRunner.run()` will be returned.
+    """
+
+    class KillWrapper(cls):
+        def __init__(self, *args, **kwargs):
+            self.armed = kwargs.pop("armed")
+            self.introduced_timeout = kwargs.pop("introduced_timeout")
+            self.terminate_time = kwargs.pop("terminate_time")
+            kill_time = kwargs.pop("kill_time")
+            self.kill_time = (
+                ((self.terminate_time or 0) + kill_time)
+                if kill_time is not None
+                else kill_time
+            )
+
+            self.process: subprocess.Popen | None = None
+            self.return_code: int | None = None
+            self.kill_counter: int = 0
+
+            super().__init__(*args, **kwargs)
+
+        def arm(self) -> None:
+            self.kill_counter = 0
+            self.armed = True
+
+        def connection_made(self, process: subprocess.Popen) -> None:
+            self.process = process
+            super().connection_made(process)
+
+        def timeout(self, fd: int | None) -> bool:
+            if self.armed and fd is None:
+                self.kill_counter += 1
+                if self.kill_time and self.kill_counter >= self.kill_time:
+                    self.process.kill()
+                    self.kill_time = None
+                if self.terminate_time and self.kill_counter >= self.terminate_time:
+                    self.process.terminate()
+                    self.terminate_time = None
+
+            # If we set the timeout argument due to a not-None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+        def process_exited(self) -> None:
+            self.return_code = self.process.poll()
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+    cmd: list,
+    protocol_class: type[Protocol],
+    *,
+    cwd: Path | None = None,
+    stdin: int | IO | bytes | Queue[bytes | None] | None = None,
+    timeout: float | None = None,
+    closing_action: Callable | None = None,
+    terminate_time: int | None = None,
+    kill_time: int | None = None,
+    **protocol_kwargs,
```

Comment:
https://github.com/datalad/datalad-next/pull/484#discussion_r1367971873 made me suspicious re whether this is a good idea. I do not see how/why the keyword arguments of `run()` would be guaranteed to be non-conflicting with kwargs of an arbitrary protocol implementation.

If they are indeed not, there should be a strong incentive for claiming that namespace in this way. Any protocol would need to obey this space, and any future addition to the `run()` API would need to be conscious of any protocol implementation out there.

Reply:
Very good point. You are right that there is potential for conflict. Will change that.

Reply:
Fixed in 25bf3cfc9cf2cbab0b502836b0aab0ff6f93edc7

That brought up the question of whether we should prefix keyword arguments of `datalad`-provided protocols with a `dl_` and document that we reserve the `dl_`* keyword space in protocols in order to prevent name clashes with external protocol developers.